packages/core/src/context/experiment/message-context.tsx

```diff
@@ -0,0 +1,96 @@
+import produce, { enableMapSet } from 'immer'
+import { ReactNode, createContext, useContext, useMemo, useState } from 'react'
+
+/**
+ * To use messages, wrap the UI components in a MessageProvider. UI components (e.g. DataPoints and InputModel) can
+ * be given an id, and messages can be shown in these components by using the given ids with the useMessageController
+ * hook.
+ *
+ * Example - setting an error message on InputModel:
+ *
+ * <InputModel id='input-model'>...
+ *
+ * const { setMessage } = useMessageController()
+ * setMessage('input-model', {
+ *    text: 'Input model cannot have duplicate ids.',
+ *    type: 'error
+ * })
+ */
+
+export type Message = {
+  type: 'info' | 'warning' | 'error' | 'custom'
+  text: string
+  disabled?: boolean
+  customComponent?: ReactNode
+}
+
+const MessageContext = createContext<
+  | {
+      messages: Map<string, Message[]>
+      setMessage: (id: string, message: Message) => void
+      setMessages: (messages: Map<string, Message[]>) => void
+      clearMessages: () => void
+    }
+  | undefined
+>(undefined)
+
+export const useMessages = (id: string | undefined) => {
+  const context = useContext(MessageContext)
+  if (context === undefined || id === undefined) {
+    return {
+      messages: undefined,
+    }
+  }
+  return {
+    messages: context.messages.get(id) ?? [],
+  }
+}
+
+export const useMessageController = () => {
+  const context = useContext(MessageContext)
+  if (context === undefined) {
+    throw new Error('useMessageController must be used within MessageProvider')
+  }
+  return {
+    setMessage: context.setMessage,
+    setMessages: context.setMessages,
+    clearMessages: context.clearMessages,
```

Comment:
Maybe it could be useful with a clearMessage method too (singular)

Reply:
True. We'd need a way to identify the message - maybe an id. I'll leave it for now but could look into it while adding disable/enable help messages.